platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/Style.java

```diff
@@ -0,0 +1,747 @@
+     * callback.
+     * </p>
+     * If the style fails to load or an invalid style URL is set, the map view will become blank.
+     * An error message will be logged in the Android logcat and {@link MapView.OnDidFailLoadingMapListener} callback
+     * will be triggered.
+     *
+     * @param url The URL of the map style
+     * @return this
+     * @see Style
+     */
+    public Builder fromUrl(String url) {
+      this.styleUrl = url;
+      return this;
+    }
+
+    /**
+     * Will load a new map style from a json string.
+     * <p>
+     * If the style fails to load or an invalid style URL is set, the map view will become blank.
+     * An error message will be logged in the Android logcat and {@link MapView.OnDidFailLoadingMapListener} callback
+     * will be triggered.
+     * </p>
+     *
+     * @return this
+     */
+    public Builder fromJson(String styleJson) {
+      this.styleJson = styleJson;
+      return this;
+    }
+
+    /**
+     * Will add the source when map style has loaded.
+     *
+     * @param source the source to add
+     * @return this
+     */
+    public Builder withSource(Source source) {
+      sources.add(source);
+      return this;
+    }
+
+    /**
+     * Will add the layer when the style has loaded.
+     *
+     * @param layer the layer to be added
+     * @return this
+     */
+    public Builder withLayer(Layer layer) {
+      layers.add(new LayerWrapper(layer));
+      return this;
+    }
+
+    /**
+     * Will add the layer when the style has loaded at a specified index.
+     *
+     * @param layer the layer to be added
+     * @return this
+     */
+    public Builder withLayerAt(Layer layer, int index) {
+      layers.add(new LayerAtWrapper(layer, index));
+      return this;
+    }
+
+    /**
+     * Will add the layer when the style has loaded above a specified layer id.
+     *
+     * @param layer the layer to be added
+     * @return this
+     */
+    public Builder withLayerAbove(Layer layer, String aboveLayerId) {
+      layers.add(new LayerAboveWrapper(layer, aboveLayerId));
+      return this;
+    }
+
+    /**
+     * Will add the layer when the style has loaded below a specified layer id.
+     *
+     * @param layer the layer to be added
+     * @return this
+     */
+    public Builder withLayerBelow(Layer layer, String belowLayerId) {
+      layers.add(new LayerBelowWrapper(layer, belowLayerId));
+      return this;
+    }
+
+    /**
+     * Will add the transition when the map style has loaded.
+     *
+     * @param transition the transition to be added
+     * @return this
+     */
+    public Builder withTransition(TransitionOptions transition) {
+      this.transitionOptions = transition;
+      return this;
+    }
+
+    /**
+     * Will add the image when the map style has loaded.
+     *
+     * @param id    the id for the image
+     * @param image the image to be added
+     * @return this
+     */
+    public Builder withImage(String id, Bitmap image) {
```

Comment:
We are missing `withImages` method.

Reply:
Atm we also don't support withLayers or withSources either but could make sense.

Reply:
Maybe postponing this into a follow up ticket as well? trying to cut some corners to merge this before our branch cut.

Reply:
Sounds good.